packages/react-utilities/src/compose/forwardRef.ts

```diff
@@ -0,0 +1,21 @@
+import * as React from 'react';
+
+/**
+ * Wrapper for `React.forwardRef`, with an API that works with components that use `IntrinsicShorthandProps`
+ * from this package. It can also be used as a cleaner API for defining any component that forwards refs.
+ *
+ * Expects the Props type to already have a `ref` attribute (which is the case if using `IntrinsicShorthandProps`).
+ * If not, add the ref attribute using `React.RefAttributes`:
+ * ```
+ * const Example = forwardRef<ExampleProps & React.RefAttributes<HTMLElement>>((props, ref) => { ... });
```

Comment:
Do you want to encourage callers to have ExampleProps be intersected with the ref attributes rather than doing it within the forwardRef call?

Reply:
I'm not 100% sure I understand, but is the suggestion that this function should add a `ref` attribute if it doesn't exist?

If the Props param doesn't have a `ref` attribute, this function can't add it because it doesn't know what type the ref should be: `Ref<HTMLDivElement>` or `Ref<HTMLSpanElement>`, etc. (It's not correct to just use `Ref<HTMLElement>`, since that would allow you to assign e.g. `Ref<HTMLButtonElement>` to a component whose root type is `<input>`.)

The Props param to this function is required to have a `ref` prop, and this example was demonstrating how to add it if it's not already there.